### PR TITLE
feat: Add support for buy token affiliate fees

### PR DESCRIPTION
**Pull request by mzhu25**
# Description

# Testing Instructions

# Checklist

-   [x] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -128,7 +128,7 @@
     },
     "dependencies": {
         "@0x/assert": "^3.0.4",
-        "@0x/asset-swapper": "0xProject/gitpkg-registry#0x-asset-swapper-v4.6.0-2989abd5d",
+        "@0x/asset-swapper": "file:/Users/michaelzhu/monorepo/packages/asset-swapper",
```
Comment (mzhu25) [inline package.json:131]: will update once the AS PR is merged and gitpkg is published

File: src/constants.ts [hunk 2]
```diff
@@ -40,6 +40,7 @@ export const PERCENTAGE_SIG_DIGITS = 4;
 export const PROTOCOL_FEE_UTILS_POLLING_INTERVAL_IN_MS = 6000;
 export const UNWRAP_QUOTE_GAS = new BigNumber(60000);
 export const WRAP_QUOTE_GAS = UNWRAP_QUOTE_GAS;
+export const AFFILIATE_FEE_TRANSFORMER_GAS = new BigNumber(10000);
```
Comment (dekz) [inline src/constants.ts:43]: probably would be a bit more than 10k?

File: src/services/swap_service.ts [hunk 3]
```diff
@@ -479,10 +492,11 @@ export class SwapService {
                 assetSwapperOpts,
             );
         } else if (buyAmount !== undefined) {
+            const buyAmountScaled = buyAmount.dividedToIntegerBy(1 - affiliateFee.buyTokenPercentFee);
```
Comment (dekz) [inline src/services/swap_service.ts:495]: hmm I was thinking 
```suggestion
            const buyAmountScaled = buyAmount.plus(buyAmount.times(affiliateFee.buyTokenPercentFee))
```

So a 100 buy 50% fee would
```
100 + 100 * 0.5 = 150 
```

versus
```
100 / (1 - 0.5) = 200 
```

unless I'm reading this wrong?

File: src/services/swap_service.ts [hunk 4]
```diff
@@ -542,25 +558,31 @@ export class SwapService {
...
         const unitMakerAssetAmount = Web3Wrapper.toUnitAmount(makerAssetAmount, buyTokenDecimals);
-        const unitTakerAssetAMount = Web3Wrapper.toUnitAmount(totalTakerAssetAmount, sellTokenDecimals);
+        const unitTakerAssetAmount = Web3Wrapper.toUnitAmount(totalTakerAssetAmount, sellTokenDecimals);
         // Best price
         const price =
             buyAmount === undefined
-                ? unitMakerAssetAmount.dividedBy(unitTakerAssetAMount).decimalPlaces(sellTokenDecimals)
-                : unitTakerAssetAMount.dividedBy(unitMakerAssetAmount).decimalPlaces(buyTokenDecimals);
+                ? unitMakerAssetAmount
+                      .times(ONE.minus(affiliateFee.buyTokenPercentFee))
```
Comment (dorothy-zbornak) [inline src/services/swap_service.ts:566]: `M / (f + 1)` if we go with the other buy scaling.

File: src/constants.ts (hunk 2 above)
Comment (mzhu25) [reply src/constants.ts:43]: probably closer to 20k if the recipient doesn't have a balance yet, but they probably do –– idk @dorothy-zbornak does 10k sound reasonable here?

File: src/handlers/swap_handlers.ts [hunk 5]
```diff
@@ -298,6 +310,39 @@ const parseGetSwapQuoteRequestParams = (
         ]);
     }
 
+    const feeRecipient = req.query.feeRecipient as string;
+    const sellTokenPercentFee = Number.parseFloat(req.query.sellTokenPercentFee as string) || 0;
+    const buyTokenPercentFee = Number.parseFloat(req.query.buyTokenPercentFee as string) || 0;
+    if (sellTokenPercentFee > 0) {
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply src/handlers/swap_handlers.ts:316]: I tested with slippagePercentage, the schema validation handles it I think

File: src/constants.ts (hunk 2 above)
Comment (dorothy-zbornak) [reply src/constants.ts:43]: Let's be conservative and go 15k for USDC.

File: src/services/swap_service.ts (hunk 3 above)
Comment (mzhu25) [reply src/services/swap_service.ts:495]: ok yeah

**Comment by mzhu25** [comment]
deploy staging

**Comment by mzhu25** [comment]
deploy staging

File: src/services/swap_service.ts [hunk 6]
```diff
@@ -479,10 +492,13 @@ export class SwapService {
                 assetSwapperOpts,
             );
         } else if (buyAmount !== undefined) {
+            const buyAmountScaled = buyAmount
+                .times(affiliateFee.buyTokenPercentageFee + 1)
+                .integerValue(BigNumber.ROUND_DOWN);
```
Comment (dorothy-zbornak) [inline src/services/swap_service.ts:497]: I feel like this should be `ROUND_UP`.

File: src/utils/service_utils.ts [hunk 7]
```diff
@@ -206,4 +208,15 @@ export const serviceUtils = {
             gasTokenGasCost,
         };
     },
+    getAffiliateFeeAmounts(quote: SwapQuote, fee: PercentageFee): AffiliateFeeAmounts {
+        const buyTokenFeeAmount = quote.worstCaseQuoteInfo.makerAssetAmount
+            .times(fee.buyTokenPercentageFee)
+            .dividedBy(fee.buyTokenPercentageFee + 1)
+            .integerValue(BigNumber.ROUND_DOWN);
```
Comment (dorothy-zbornak) [inline src/utils/service_utils.ts:215]: I like that the the fee is only off the guaranteed price.

File: src/services/swap_service.ts (hunk 6 above)
Comment (mzhu25) [reply src/services/swap_service.ts:497]: I believe the rounding mode needs to be the same between this and `getAffiliateFeeAmounts`

**Comment by mzhu25** [comment]
deploy staging

File: src/services/swap_service.ts (hunk 6 above)
Comment (dorothy-zbornak) [reply src/services/swap_service.ts:497]: Isn't it possible we'll be under by 1 wei after the affiliate fee is taken out by the transformer?

Comment (mzhu25) [reply src/services/swap_service.ts:497]: I tried to work out the math but gave up
```
    for (let i = 0; i < 100000; i++) {
        it(`Test ${i}`, () => {
            const buyAmount = getRandomInteger(0, constants.ONE_ETHER);
            const buyTokenPercentageFee = getRandomFloat(0, 1);
            const buyAmountScaled = buyAmount.times(buyTokenPercentageFee.plus(1)).integerValue(BigNumber.ROUND_DOWN);
            const buyTokenFeeAmount = buyAmountScaled
                .times(buyTokenPercentageFee)
                .dividedBy(buyTokenPercentageFee.plus(1))
                .integerValue(BigNumber.ROUND_DOWN);
            expect(buyAmountScaled.minus(buyTokenFeeAmount)).to.bignumber.equal(buyAmount);
        });
    }
```
> `100000 passing (26s)`

**Comment by mzhu25** [comment]
deploy production

**Pull request merged**